lib/linked_list.rb

```diff
@@ -12,152 +12,320 @@ def initialize(value, next_node = nil)
+      @head = @head.next
+      return
     end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
-    def reverse
-      raise NotImplementedError
+    current = @head
+    while !current.next.nil?
+      if current.next.data == value
+        current.next = current.next.next
+        return
+      end
+      current = current.next
     end
+    return
+  end
 
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n)
+  # Space Complexity O(1)
+  def reverse
+    current_node = @head
+    prev_node = nil
+    next_node = nil
+    while !current_node.nil?
+      next_node = current_node.next
+      current_node.next = prev_node
+      prev_node = current_node
+      current_node = next_node
+    end
+    @head = prev_node
+  end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def find_middle_value
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_middle_value
+    return nil if @head.nil?
+    current = @head
+    len = self.length
+    return @head if len == 1
+    mid_val = len / 2
+    count = 1
+    while count <= len
+      if count == mid_val
+        return current.data
+      end
+      current = current.next
+      count += 1
     end
+  end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
-    def find_nth_from_end(n)
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity:O(n)
+  # Space Complexity:O(1)
+  def find_nth_from_end(n)
+    current = @head
+    len = self.length
+    count = len - 1
+    return if n > count
+    current = @head
+    while !current.nil?
+      if count == n
+        return current.data
+      end
+      current = current.next
+      count -= 1
     end
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def has_cycle
+    visited_nodes = []
+    current = @head
+    while !current.nil?
+      if visited_nodes.include?(current)

```

Comment:
Because `.include?` traverses through `visited_nodes`, this is **another** loop!  So your time complexity is actually O(n<sup>2</sup>).

Is there another way to do this without a 2ndary array?